client/client-lib/src/api.rs

```diff
@@ -1,148 +1,615 @@
+                }
+            }));
+        }
 
-    /// Upload ecash (encrypted) backup for mint to safekeep
-    async fn upload_ecash_backup(&self, request: &fedimint_mint::SignedBackupRequest)
-        -> Result<()>;
+        let mut member_delay_ms = BTreeMap::new();
+        let mut member_errors = BTreeMap::new();
 
-    /// Download ecash (encrypted) backup from mint to safekeep
-    async fn download_ecash_backup(
+        // Delegates the response handling to the `QueryStrategy` with an exponential back-off
+        // with every new set of requests
+        let max_delay_ms = 1000;
+        loop {
+            let response = futures.next().await;
+            trace!(?response, method, ?params, "Received member response");
+            match response {
+                Some(PeerResponse { peer, result }) => {
+                    let result: Result<MemberRet> = result.map_err(ApiError::Rpc).and_then(|o| {
+                        serde_json::from_value::<MemberRet>(o)
+                            .map_err(|e| ApiError::ResponseDeserialization(e.into()))
+                    });
+
+                    let strategy_step = strategy.process(peer, result);
+                    trace!(
+                        method,
+                        ?params,
+                        ?strategy_step,
+                        "Taking strategy step to the response after member response"
+                    );
+                    match strategy_step {
+                        QueryStep::RetryMembers(peers) => {
+                            for retry_peer in peers {
+                                member_errors.remove(&retry_peer);
+
+                                let mut delay_ms =
+                                    member_delay_ms.get(&retry_peer).copied().unwrap_or(10);
+                                delay_ms = cmp::min(max_delay_ms, delay_ms * 2);
+                                member_delay_ms.insert(retry_peer, delay_ms);
+
+                                futures.push(Box::pin({
+                                    let method = &method;
+                                    let params = &params;
+                                    async move {
+                                        // Note: we need to sleep inside the retrying future,
+                                        // so that `futures` is being polled continously
+                                        sleep(Duration::from_millis(delay_ms)).await;
+                                        PeerResponse {
+                                            peer: retry_peer,
+                                            result: self
+                                                .request_raw(retry_peer, method, params)
+                                                .await,
+                                        }
+                                    }
+                                }));
+                            }
+                        }
```

Comment:
This can potentially lead to multiple parallel requests to the same peer from different retries and could lead to an exponential number of parallel reqeuests. Maybe we should remove pending requests to the peers to which we start new ones?

Reply:
That would only happen if the strategy returned `Retry` for the same peer, multiple times, before getting response from it first. Which is kind of a bug.

It would be better to remove the previous future, and it might even be useful for certain strategies, but I don't see an easy way to do https://docs.rs/futures/latest/futures/prelude/stream/struct.FuturesUnordered.html ATM.

Reply:
> Would have to write own `FutureUnordered`-like thing or look for something.

I don't think so. [`select_all`](https://docs.rs/futures/latest/futures/future/fn.select_all.html) should do exactly what we require here: wait for any future to complete and give back a vec of the remaining ones.

Reply:
@elsirion It might be the primitive behind `FutureUnordered` but then we would need to track and remove them as well, also wrap each in some `FutureWithId` ... 

... it's doable, I'm sure. :)